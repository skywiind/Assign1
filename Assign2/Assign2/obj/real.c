#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL{
    double value;
};

REAL* newREAL(double x){
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
}

double getREAL(REAL *v){
    return v->value;
}

double setREAL(REAL *v,double x){
    double old = v->value;
    v->value = x;
    return old;
}

void displayREAL(void *v,FILE *fp){
    fprintf(fp,"%lf",getREAL((REAL*) v));
	return;
}

double compareREAL(void *v,void *w){
    return getREAL(v) - getREAL(w);
}

void freeREAL(void *v){
    free((REAL *) v);
	return;
}
