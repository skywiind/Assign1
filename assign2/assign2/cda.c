/*****************************************************************************************
*                                           *                                            *
*             CS201-001 Assign 0            * This file contains the implementations of  *
*             Wyatt Fairbanks               * the methods for the circular dynamic       *
*             Dr. John Lusth                * array class.                               *
*             cda.c                         *                                            *
*             8/21/2018                     *                                            *
*                                           *                                            *
*****************************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "cda.h"

/*Class definition for CDA. Declares all internal data structures, counters, and
  function pointers.*/
struct cda {
	void **arr;
	int    size;
	int    cap;
	int    debug;
	int    start;
	void(*display)(void *value, FILE *fp);
	void(*free)(void *value);
};

/*Constructor for CDA. Allocates memory and initializes everything. Returns the new CDA.*/
CDA *newCDA(void) {
	CDA *items = malloc(sizeof(CDA));
	assert(items != 0);
	items->arr = malloc(sizeof(void *));
	items->size = 0;
	items->cap = 1;
	items->debug = 0;
	items->start = 0;
	items->display = 0;
	items->free = 0;
	return items;
}

/*Method to set the siplay function for the underlying data type of the CDA.*/
void setCDAdisplay(CDA *items, void(*display)(void *value, FILE *fp)) {
	items->display = display;
	return;
}

/*Method to set the free function for the underlying data type of the CDA.*/
void setCDAfree(CDA *items, void(*free)(void *value)) {
	items->free = free;
	return;
}

/*Internal utility function. Changes a user-scope index to the internal scope.*/
int indexCDA(CDA *items, int index) {
	return ((items->start + index) + items->cap) % items->cap;
}

/*Method to manage memory, expand or contracting as necessary.*/
void memCDA(CDA *items) {
	if (items->size == items->cap) {
		void **temp = items->arr;
		items->arr = malloc((sizeof(void *)) * (items->cap * 2));
		for (int i = 0; i < items->size; i++) {
			items->arr[i] = temp[indexCDA(items, i)];
		}
		items->start = 0;
		items->cap *= 2;
		free(temp);
		return;
	}
	if ((items->size * 4) < items->cap) {
		void **temp = items->arr;
		items->arr = malloc((sizeof(void*)) * (items->cap / 2));
		for (int i = 0; i < items->size; i++) {
			items->arr[i] = temp[indexCDA(items, i)];
		}
		items->start = 0;
		items->cap /= 2;
		free(temp);
		return;
	}
	return;
}

/*Method to insert new items into CDA. Manages memory using a temp array,
  inserts, shifting based on efficiency.*/
void insertCDA(CDA *items, int index, void *value) {
	//Manage memory
	if (items->size == items->cap) {
		memCDA(items);
	}

	//Insert at front
	if (index == 0) {
		//Insert first item
		if (items->size == 0) {
			items->arr[0] = value;
			items->size++;
			return;
		}
		items->arr[indexCDA(items, -1)] = value;
		items->start = indexCDA(items, -1);
		items->size++;
		return;
	}

	//Insert at back
	else if (index == items->size) {
		items->arr[indexCDA(items, items->size)] = value;
		items->size++;
		return;
	}
	else {
		//Insert right of center
		if (index > items->size / 2) {
			for (int i = items->size; i > index; i--) {
				items->arr[indexCDA(items, i)] = items->arr[indexCDA(items, i - 1)];
			}
			items->arr[indexCDA(items, index)] = value;
		}
		//Insert left of center
		else if (index <= items->size / 2) {
			for (int i = -1; i < index - 1; i++) {
				items->arr[indexCDA(items, i)] = items->arr[indexCDA(items, i + 1)];
			}
			items->arr[indexCDA(items, index - 1)] = value;
			items->start = (indexCDA(items, -1));
		}
		items->size++;
	}
	return;
}

/*Method to remove items from the CDA. Manages memory using a temp array,
  removes, shifting based on efficiency. Returns the removed item.*/
void *removeCDA(CDA *items, int index) {

	assert(index >= 0);
	if (index > items->size) {
		return NULL;
	}
	void *removed = items->arr[indexCDA(items, index)];

	//Remove at front
	if (index == 0) {
		items->start = indexCDA(items, 1);
	}
	
	else if (index > 0 && index < (items->size - 1)) {
		//Remove right of center
		if (index > items->size / 2) {
			for (int i = index; i < items->size - 1; i++) { 
				items->arr[indexCDA(items, i)] = items->arr[indexCDA(items, i + 1)];
			}
		}
		//Remove left of center
		else if (index <= items->size / 2) {
			for (int i = index; i > 0; i--) { 
				items->arr[indexCDA(items, i)] = items->arr[indexCDA(items, i - 1)];
			}
			items->start = indexCDA(items, 1);
		}
	}
	items->size--;
	
	//Manage memory
	if (items->size == 0) {
		items->arr = realloc(items->arr, sizeof(void *));
		items->cap = 1;
		items->start = 0;
	}
	else if ((items->size * 4) < items->cap) {
		memCDA(items);
	}
	return removed;
}

/*Method for joing two CDAs. Uses a temp array and repeated insertion to 
  add donor array's contents to the recipient.*/
void unionCDA(CDA *recipient, CDA *donor) {
	void **temp = donor->arr;
	for (int i = 0; i < donor->size; i++) {
		insertCDAback(recipient, donor->arr[indexCDA(donor, i)]);
	}
	donor->size = 0;
	donor->cap = 1;
	donor->arr = malloc(sizeof(void *));
	free(temp);
	return;
}

/*Method for retriving value stored in CDA at target index.
  Returns the targeted value.*/
void *getCDA(CDA *items, int index) {
	assert(index >= 0 && index < items->size);
	return items->arr[indexCDA(items, index)];
}


/*Method for replacing a value in the CDA at the target index. If empty,
  value is inserted. If not, original value is replaced and returned.*/
void *setCDA(CDA *items, int index, void *value) {
	assert(index >= -1 && index <= items->size);
	if (index == -1) {
		insertCDAfront(items, value);
		return 0;
	}
	if (index == items->size) {
		insertCDAback(items, value);
		return 0;
	}
	void *temp = items->arr[indexCDA(items, index)];
	items->arr[indexCDA(items, index)] = value;
	return temp;
}

/*Method for retrieving the size of the CDA. Returns the size.*/
int sizeCDA(CDA *items) {
	return items->size;
}

/*Method for displaying the contents of the CDA. Uses the debug flag and
  display function pointer to decide display method.*/
void displayCDA(CDA *items, FILE *fp) {
	fprintf(fp, "(");
	if (items->size == 0 && items->debug > 0) {
		fprintf(fp, "(%d)", ((items->cap) - (items->size)));
	}
	for (int i = 0; i < items->size; i++) {
		if (items->display == 0) {
			fprintf(fp, "@%p", items->arr[indexCDA(items, i)]);
		}
		else {
			items->display(items->arr[indexCDA(items, i)], fp);
		}
		if (items->debug > 0 && i == (items->size - 1)) {
			fprintf(fp, ",(%d)", ((items->cap) - (items->size)));
		}
		if (i != (items->size) - 1) {
			fprintf(fp, ",");
		}
	}
	fprintf(fp, ")");
	return;
}

/*Method for setting the CDA debug flag to the specified value. Returns original value.*/
int debugCDA(CDA *items, int level) {
	int prev = items->debug;
	items->debug = level;
	return prev;
}

/*Method for freeing the memory of the CDA. Calls the set free function for each
  item in the array. If no functions if set, it simply frees the whole array.*/
void freeCDA(CDA *items) {
	if (items->free != 0) {
		for (int i = 0; i < items->size; i++) {
			items->free(items->arr[indexCDA(items, i)]);
		}
	}
	free(items->arr);
	free(items);
	return;
}