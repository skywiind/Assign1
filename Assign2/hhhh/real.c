#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL{
    double value;
};

REAL* newREAL(double x){
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
}

double getREAL(REAL *v){
    return v->value;
}

double setREAL(REAL *v,double x){
    double old = v->value;
    v->value = x;
    return old;
}

void displayREAL(void *v,FILE *fp){
    fprintf(fp,"%lf",getREAL((REAL*) v));
	return;
}

int compareREAL(void *v,void *w){
    double r = getREAL(v) - getREAL(w);
    if (r < 0.0) {
	    return -1;
    } else if (r > 0.0) {
	    return 1;
    } else {
	    return 0;
    }
}

void freeREAL(void *v){
    free((REAL *) v);
	return;
}
