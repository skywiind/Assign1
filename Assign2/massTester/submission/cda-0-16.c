#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "integer.h"
#include "real.h"
#include "string.h"
#include "cda.h"

void srandom(unsigned int);
long int random(void);

static void
display(CDA *p,char *intro)
    {
    printf("%s",intro);
    if (sizeCDA(p) < 200)
        {
        displayCDA(p,stdout);
        }
    else
        {
        int k = 100;
        int j;
        printf("(displaying first and last %d elements)\n",k);
        for (j = 0; j < k; ++j)
            {
            displayINTEGER(getCDA(p,j),stdout);
            printf(" ");
            }
        printf("... ");
        for (j = 0; j < k; ++j)
            {
            displayINTEGER(getCDA(p,sizeCDA(p)-k+j),stdout);
            if (j < k - 1) printf(" ");
            }
        printf("\n");
        }
    printf("\n");
    }

int
main(void)
    {
    srandom(28);
    int i,r,rounds=2,size=1964;
    printf("INTEGER test of CDA, unions\n");
    printf("test size is %d\n",size);
    CDA *p = newCDA();
    setCDAdisplay(p,displayINTEGER);
    setCDAfree(p,freeINTEGER);
    debugCDA(p,1);
    CDA *q = newCDA();
    setCDAdisplay(q,displayINTEGER);
    setCDAfree(q,freeINTEGER);
    debugCDA(q,1);
    display(p,"p items: ");
    display(q,"p items: ");
    for (r = 0; r < rounds; ++r)
        {
        int j,k;
        for (i = 0; i < size; ++i)
            {
            j = random() % size;
            k = random() % (sizeCDA(p) + 1);
            insertCDA(p,k,newINTEGER(j));
            j = random() % size;
            k = random() % (sizeCDA(q) + 1);
            insertCDA(q,k,newINTEGER(j));
            }
        for (i = 0; i < size - random() % size / 2; ++i)
            {
            unionCDA(p,q);
            unionCDA(q,p);
            }
        display(p,"p items: ");
        display(q,"q items: ");
        printf("p size is %d\n",sizeCDA(p));
        printf("q size is %d\n",sizeCDA(q));
        }
    freeCDA(p);
    freeCDA(q);
    return 0;
    }
