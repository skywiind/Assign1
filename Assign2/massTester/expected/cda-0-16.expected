INTEGER test of CDA, unions
test size is 1964
p items: ((1))
p items: ((1))
p items: ((1))
q items: (displaying first and last 100 elements)
721 1744 1494 1590 1247 1558 969 708 85 1671 213 1693 433 340 1084 1804 355 584 95 962 388 468 1076 1853 1142 641 1757 1413 1219 368 1576 1282 1018 849 1730 367 1784 867 575 601 1238 884 160 1693 1374 1580 275 1850 1252 1917 1471 1687 821 96 979 413 886 1829 586 423 1536 547 1274 1196 659 1640 1304 658 1227 146 29 1369 1699 36 1854 947 464 424 1106 1357 328 450 1268 1078 1592 1701 285 852 1594 1816 28 586 753 549 1960 119 1739 1601 183 1320 ... 11 249 442 1310 1287 1007 1720 1218 1163 1764 989 1411 1880 1476 1186 757 1266 1012 1385 787 476 1491 261 31 1180 725 1616 1360 1303 1073 755 1185 1285 1121 493 836 752 975 1102 773 18 1850 1331 89 1909 826 70 992 1501 1564 1588 60 492 800 1326 1391 1049 1304 1074 437 1409 653 738 400 1865 1369 1859 760 1566 176 702 1284 1859 1424 364 685 1384 1724 1016 1430 845 1496 14 1700 74 1114 1097 530 1169 328 1607 919 16 730 53 264 1068 1009 427 451

p size is 0
q size is 3928
p items: ((1))
q items: (displaying first and last 100 elements)
452 1564 1514 1728 647 1838 1424 1196 1398 1880 265 1227 895 1013 1480 1441 1682 404 1409 1426 263 672 921 14 1702 144 1648 285 757 487 1891 1628 1841 1623 1083 1625 230 1067 984 1763 617 947 564 765 746 1840 1429 461 1116 1461 1509 805 1034 1839 274 143 37 1090 1434 785 1171 1874 1614 1822 1049 372 1521 57 779 32 682 1047 1955 97 497 1066 1723 727 18 1356 124 1840 907 344 457 1857 473 958 385 859 1075 1784 1598 798 606 1292 1191 846 1198 99 ... 1102 463 773 611 241 360 18 669 1482 1850 1331 89 1909 1152 826 70 206 1302 1928 1257 411 992 1501 1564 1588 60 492 800 1326 1391 1049 1304 902 1074 437 580 1409 1316 653 738 400 1865 1369 1859 217 760 1306 1566 891 176 377 702 1284 1748 1859 423 1424 364 685 1540 1384 140 965 1317 1724 1016 1430 1681 845 1496 14 337 1700 74 1114 1097 1638 530 1476 1169 328 1607 1228 1769 919 1279 16 730 53 264 1068 1009 44 1499 974 9 1221 427 1181 451

p size is 0
q size is 7856
